packages/amplify-graphql-schema-generator/src/ts-schema-generator/helpers.ts

```diff
@@ -70,14 +83,21 @@ const createDataType = (type: FieldType): ts.Node => {
   }
 
   // List or NonNull
-  const modifier = type.kind === 'List' ? TYPESCRIPT_DATA_SCHEMA_CONSTANTS.ARRAY_METHOD : TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REQUIRED_METHOD;
+  const modifier =
+    field.type.kind === 'List' ? TYPESCRIPT_DATA_SCHEMA_CONSTANTS.ARRAY_METHOD : TYPESCRIPT_DATA_SCHEMA_CONSTANTS.REQUIRED_METHOD;
+  const unwrappedField = new Field(field.name, field.type.type);
   return ts.factory.createCallExpression(
-    ts.factory.createPropertyAccessExpression(createDataType(type.type) as ts.Expression, ts.factory.createIdentifier(modifier)),
+    ts.factory.createPropertyAccessExpression(createDataType(unwrappedField) as ts.Expression, ts.factory.createIdentifier(modifier)),
     undefined,
     undefined,
   );
 };
 
+const isSequenceField = (field: Field): boolean => {
+  const sequenceRegex = /^nextval\(.+::regclass\)$/;
+  return field.default?.kind === 'DB_GENERATED' && sequenceRegex.test(field.default.value.toString());
```

Comment:
(Copying this from main conversation comment)

Can we find documentation that this is the canonical way to determine whether a field is `serial`? In the past we've relied on experimentally-verified "rules" that turned out to be incomplete, so I'd love to get ahead of that this time.

Reply:
In summary we cannot determine if a column was declared `serial`, we _can_ identify sequences though, and from this, infer serial fields.

Ultimately `serial` doesn't really exist as a datatype we can directly introspect. In PG it is basically a macro for _"Create a column as an integer and create a sequence to back that column as a default"_. We end up with an ordinary non-null integer with a special default value.

We can tell if a column is backed by a _sequence_ since its default will be `nextval(_)`, one of the special [sequence functions](https://www.postgresql.org/docs/9.2/functions-sequence.html).

[Reference](https://dba.stackexchange.com/questions/47098/how-do-i-determine-if-a-column-is-defined-as-a-serial-data-type-instead-of-an-in).

Reply:
Thanks for the thorough explanation!